### DISCUSSION TITLE
OIDC with Discord

### DISCUSSION THREAD

**Discussion by Conquerix** [outline/outline#6355] [Show and tell] [answered]
Hi,

I recently setup OIDC with discord and thought other people would be interested!

#### Quick summary of my installation:

I use NixOS for all my machines so installing outline in itself is as simple as filling the information below:

```nix
services.outline = {
  enable = cfg.enable;
  publicUrl = "https://${cfg.url}";
  port = cfg.port;
  storage.storageType = "local";
  forceHttps = false;
  oidcAuthentication = {
    authUrl = "https://discordapp.com/api/oauth2/authorize";
    tokenUrl = "https://discordapp.com/api/oauth2/token";
    userinfoUrl = "https://discordapp.com/api/users/@me";
    clientId = "1170421650861334618";
    clientSecretFile = "/secrets/outline-client-secret-key";
    scopes = [ "email" "identify" ];
    usernameClaim = "preferred_username";
    displayName = "Discord";
  };
};
```

#### Setting up discord OIDC

What interests us here is the `oidcAuthentication` part.

The `authUrl`, `tokenUrl` and `userinfoUrl` are generic URLs for anyone who would want to set this up. Same goes for the scopes, those work, I haven't tried with others though. I still don't know what the `usernameClaim` field does, but my setup works. 

To find your clientId and clientSecret, you first have to go to [Discord's Developer Portal](https://discord.com/developers/applications), in the applications category.

Once there create a new application.

Once it is done, go to the OAuth2 tab on the left:

![Discord OAuth2 tab screenshot](https://github.com/outline/outline/assets/74060335/83de9d97-3770-4617-ae6b-a2c3337dd92c)

Once there you will find your clientID and clientSecret (you may need to reset it to generate it).

Don't forget to add a callback link like I did to redirect back to your outline domain, it should be in the shape of `https://your-domain.com/auth/oidc.callback`.

And then you're all set !

**Comment by cstby** [comment, marked as answer]
Thank you @Conquerix for sharing this guide! I really think this could be helpful to have in the official documentation.

I can confirm that the above instructions produce the desired result of authentication with Discord.

For those of us unfamiliar with Nix OS, I've included the relevant lines from my .env file below. The client id and secret should be replaced with those from your discord app, but the rest can be used without any tweaks.

```
OIDC_AUTH_URI=https://discordapp.com/api/oauth2/authorize
OIDC_TOKEN_URI=https://discordapp.com/api/oauth2/token
OIDC_USERINFO_URI=https://discordapp.com/api/users/@me
OIDC_CLIENT_ID=123456
OIDC_CLIENT_SECRET=abcdefg
OIDC_SCOPES=email identify
OIDC_DISPLAY_NAME=Discord
```

**Comment by NotWorkingCode** [comment]
Hello. Now when I try to log in I get an error: Invalid key length

```
outline           | {"error":"Invalid key length","level":"error","message":"Error during authentication","stack":"RangeError: Invalid key length\n    at Cipheriv.createCipherBase (node:internal/crypto/cipher:121:19)\n    at Cipheriv.createCipherWithIV (node:internal/crypto/cipher:140:3)\n    at new Cipheriv (node:internal/crypto/cipher:243:3)\n    at Object.createCipheriv (node:crypto:146:10)\n    at Object.createCipheriv (/opt/outline/node_modules/dd-trace/packages/datadog-instrumentations/src/crypto.js:28:27)\n    at UserAuthentication.set (/opt/outline/node_modules/sequelize-encrypted/index.js:48:33)\n    at setEncryptedColumn (/opt/outline/build/server/models/decorators/Encrypted.js:52:55)\n    at set accessToken (/opt/outline/build/server/models/UserAuthentication.js:46:39)\n    at UserAuthentication.set (/opt/outline/node_modules/sequelize/lib/model.js:2256:32)\n    at UserAuthentication.set (/opt/outline/node_modules/sequelize/lib/model.js:2241:18)\n    at UserAuthentication._initValues (/opt/outline/node_modules/sequelize/lib/model.js:156:10)\n    at new Model (/opt/outline/node_modules/sequelize/lib/model.js:116:10)\n    at new Model (/opt/outline/node_modules/sequelize-typescript/dist/model/model/model.js:14:9)\n    at new IdModel (/opt/outline/build/server/models/base/IdModel.js:20:5)\n    at new UserAuthentication (/opt/outline/build/server/models/UserAuthentication.js:31:5)\n    at new user_authentication (/opt/outline/build/server/models/decorators/Fix.js:21:7)\n    at user_authentication.build (/opt/outline/node_modules/sequelize/lib/model.js:1344:12)\n    at Model.<computed> (/opt/outline/node_modules/sequelize-typescript/dist/model/model/model.js:128:28)\n    at Model.<computed> [as build] (/opt/outline/node_modules/sequelize-typescript/dist/model/model/model.js:138:28)\n    at /opt/outline/node_modules/sequelize/lib/model.js:1358:43\n    at Array.map (<anonymous>)\n    at user_authentication.bulkBuild (/opt/outline/node_modules/sequelize/lib/model.js:1358:22)"}
```

**Comment by sebasptsch** [comment]
I use Outline for several communities that I'm in, both authenticated through Discord. I've opened a PR (#7005) that adds improved Discord support like adding the server icon, name and nicknames. This also adds Discord Role whitelist for internal wikis of Discord servers.

**Reply by cstby** [reply]
Very cool! I've subscribed to the PR, and I'm hoping it gets merged.

**Reply by sebasptsch** [reply]
I am too lol

**Reply by cstby** [reply]
Looks like your PR was just merged. Congrats! Would you mind providing some instructions or documentation on how to use the new functionality?

**Reply by tommoor** [reply]
I have some draft documentation ready to publish once this is in a release, [here](https://docs.getoutline.com/s/hosting/doc/discord-g4JdWFFub6).

**Reply by sebasptsch** [reply]
That's awesome, is there a bleeding edge release channel to use so we can start to use this feature?

**Reply by tommoor** [reply]
I just published a pre-release straight off `main` – `outlinewiki/outline:0.78.0-0`

**Reply by QYG2297248353** [reply]
```env
# To configure Discord auth, you'll need to create a Discord Application at
# => https://discord.com/developers/applications/
#
# When configuring the Client ID, add a redirect URL under "OAuth2":
# https://<URL>/auth/discord.callback
DISCORD_CLIENT_ID=
DISCORD_CLIENT_SECRET=

# DISCORD_SERVER_ID should be the ID of the Discord server that Outline is
# integrated with. 
# Used to verify that the user is a member of the server as well as server
# metadata such as nicknames, server icon and name.
DISCORD_SERVER_ID=

# DISCORD_SERVER_ROLES should be a comma separated list of role IDs that are
# allowed to access Outline. If this is not set, all members of the server
# will be allowed to access Outline.
# DISCORD_SERVER_ID and DISCORD_SERVER_ROLES must be set together.
DISCORD_SERVER_ROLES=
```
from [.env.sample](https://github.com/outline/outline/blob/main/.env.sample)

May I ask if there is a temporary documentation for the relevant configurations, I'm not sure if it's enabled in `0.77.3`

**Reply by sebasptsch** [reply]
It'll be released in 0.78 afaik

**Reply by QYG2297248353** [reply]
Okay, I'll be patient.

**Reply by tommoor** [reply]
https://github.com/outline/outline/releases/tag/v0.78.0 is now published.
